### ISSUE TITLE
JS execution for an invisible element

### ISSUE THREAD

**Issue by denis-anisimov** [vaadin/flow#3211]
Technically it's not clear what should happen with the JS execution for an invisible element.

Two aspects: 
- on the server side
- on the client side.

There is a request to avoid server-side updates for invisible element.
JS execution is kind of server side update. Should it be also skipped ?

Questions are: 
- totally ignored (and discarded)
- deferred 

In any case server side behavior needs correction regarding this.

If we do something on the server side then the client side most likely doesn't need any modifications.

But if we still send JS executions to the client side then we also need to do something there.

At least the initially invisible element won't be bound (#3148). It means that JS execution may be invalid on the not entirely initialized `StateNode`. 
I'm going to defer JS execution for initially disabled element as a part of the implementation of #3148.

Nothing is done for JS execution if the element has been initially visible then became invisible.
JS execution requests will be executed for such element in a regular way.
May be this should be changed.

This ticket is rather a number of questions than ticket to implement.
We need to agree what we do with JS execution for invisible element and either put answers here and make this ticket ready to be worked on or create a number of separate tickets.

**Comment by pleku** [comment]
I think we should keep things simple and not allow JS execution for invisible components (and subparts). I would be fine for even just throwing for that on the server side to make it really clear for users.

If we get feedback that this should be needed, then we can have another look at this.

**Comment by denis-anisimov** [comment]
One issue with this.

I don't think we can do this actually.

The example is: `Grid`.
It executes JS in its CTOR once the component is attached.
We can't prevent attaching invisible components. Otherwise a lot of thins becomes too complicated.

It means that initially invisible Grid will never be initialized correctly. So it won't just work.

`Grid` is just an example. We cannot assume that developers cares about visibility/invisibility in their custom components. They are free to call JS at any time. So they may do it in the CTOR, on component attach.

But it the component is used in 3d party code it may be set invisible.

Since the component developer relies on JS execution code we may not just simply ignore it at ll.

So looks like ignoring JS on the server side is not an option at all.

So we should either:
- defer JS execution on the server side
- defer JS execution in the client side

**Comment by pleku** [comment]
That is true, the JS execution needed for component initialization on the client side is something that should work _somehow_ regardless if the component is initially _invisible_ or not.

I still consider that the component user's execution of JS for invisible components is something that we don't have to support. But probably the most sophisticated solution would be to fix this in a manner that handles both needs.

**Comment by denis-anisimov** [comment]
You always forget one important thing: there is no such concept as JS execution for the _node_.

We have a generic JS execution concept. In the end all JS executions are just delegated to the `Page::executeJavaScript`.
And a node may be just an argument there. So it's about checking whether some _argument_ is visible or not.......
So it's not easy at all.

On the client side the parameters are "parsed" to find the elements (and their `StateNode`s) and we already have there some logic which defers the JS execution in some cases. E.g. for virtual children JS execution is deferred because of their possible async nature. So on the client side we may detect that element is visible or not (already).

One more mistake that I've also been doing: `setVisible()` is open API. But JS execution can be internal impl of the component and can be external request.
It's OK do skip calling JS when it's requested outside of the component (developer which uses the component asks to execute JS via existing API).
But it's not OK to skip JS execution if it's a part of component implementation.

I don't remember the whole of `Grid` impl but I'm pretty sure that there are such cases.
Let's assume that you have an API method in your component which sets some data on the server side ( like DataProvider) and then sends this data to the client side via JS.
If we skip JS execution entirely then the element becomes desynchronized : its server side data is not the same as client side data. And once the element becomes visible it will be a serious problem.
We should execute JS once the element becomes visible.
At the moment the JS will be just executed regardless of visibility.

I'm pretty sure that we can leave with the current behavior. It doesn't introduce any issues at least.
Only JS execution for the initially invisible node is deferred on the client side ( done as a part of #3212). 

So this is not critical issue. But we should be aware of it at least.

And I still think that the most appropriate way here: always send JS execution requests from the server side to the client side and defer them on the client side if the element is invisible.

**Comment by Legioth** [comment]
Would cases such as Grid help if `beforeClientReponse` would be triggered only after the element has been set as visible? That approach should hopefully protect against all proper use of `executeJavaScript`, even though some sloppy practices might probably still cause problems.

**Comment by Legioth** [comment]
Another potential approach is the same idea that was suggested in relationship with some other problems with the timing of `executeJavaScript`, namely that the caller would have some way of indicating whether a missing node reference should cause the execution to be delayed or if a `null` parameter should be passed instead.
